numba/targets/codegen.py

```diff
@@ -387,9 +389,63 @@ def get_pointer_to_function(self, name):
         return self._codegen._engine.get_function_address(name)
 
     def _finalize_specific(self):
+        self._codegen._scan_and_fix_unresolved_refs(self._final_module)
         self._codegen._engine.finalize_object()
 
 
+class RuntimeLinker(object):
+    """
+    For tracking unresolved symbols generated at runtime due to recursion.
+    """
+    PREFIX = '.numba.unresolved$'
+
+    def __init__(self):
+        self._unresolved = defaultdict(list)
+        self._defined = set()
+        self._resolved = []
+
+    def scan_unresolved_symbols(self, module, engine):
+        """
+        Scan and track all unresolved external symbols in the module and
+        allocate memory for it.
+        """
+        prefix = self.PREFIX
+
+        for gv in module.global_variables:
+            if gv.name.startswith(prefix):
+                sym = gv.name[len(prefix):]
+                # Allocate a memory space for the pointer
+                abortfn = engine.get_function_address('nrt_unresolved_abort')
+                ptr = ctypes.c_void_p(abortfn)
+                engine.add_global_mapping(gv, ctypes.addressof(ptr))
+                self._unresolved[sym].append(ptr)
+
+    def scan_defined_symbols(self, module):
+        """
+        Scan and track all defined symbols.
+        """
+        for fn in module.functions:
+            if not fn.is_declaration:
+                self._defined.add(fn.name)
+
+    def resolve(self, engine):
+        """
+        Fix unresolved symbols if there are defined.
+        """
+        # An iterator to get all unresolved but available symbols
+        pending = [name for name in self._unresolved if name in self._defined]
+        # Resolve pending symbols
+        for name in pending:
+            # Get runtime address
+            fnptr = engine.get_function_address(name)
+            # Fix all usage
+            for ptr in self._unresolved[name]:
+                ptr.value = fnptr
+                self._resolved.append((name, ptr))   # keep ptr alive
+            # Delete resolved
+            del self._unresolved[name]
```

Comment:
At the end of this function, is enough information available to abort compilation at this point if there are unresolvable symbols? Or do e.g. external refs come in to play?